src/main/java/io/cdap/plugin/WindowAggregation.java

```diff
@@ -0,0 +1,450 @@
+import io.cdap.cdap.etl.api.lineage.field.FieldTransformOperation;
+import io.cdap.plugin.function.DiscretePercentile;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.expressions.Window;
+import org.apache.spark.sql.expressions.WindowSpec;
+import org.apache.spark.sql.functions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.InvalidParameterException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Window aggregator plugin
+ */
+@Plugin(type = SparkCompute.PLUGIN_TYPE)
+@Name(WindowAggregation.NAME)
+@Description("Specify a window over which functions should be applied. \n" +
+  "Supports functions: Rank, Dense Rank, Percent Rank, N tile, Row Number, Median," +
+  "Continuous Percentile, Lead, Lag, First, Last, Cumulative distribution, Accumulate.")
+public class WindowAggregation extends SparkCompute<StructuredRecord, StructuredRecord> {
+
+  public static final String NAME = "WindowAggregation";
+  private final WindowAggregationConfig config;
+  private Schema outputSchema;
+
+  private static final Logger LOG = LoggerFactory.getLogger(WindowAggregation.class);
+
+  public WindowAggregation(WindowAggregationConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+
+    StageConfigurer stageConfigurer = pipelineConfigurer.getStageConfigurer();
+    Schema inputSchema = stageConfigurer.getInputSchema();
+
+    List<String> partitionFields = config.getPartitionFields();
+    List<WindowAggregationConfig.FunctionInfo> aggregates = config.getAggregates();
+    List<String> partitionOrderFields = config.getPartitionOrderFields();
+
+    if (inputSchema == null || partitionFields.isEmpty() || aggregates.isEmpty()) {
+      stageConfigurer.setOutputSchema(null);
+      return;
+    }
+
+    validate(inputSchema, partitionFields, aggregates, partitionOrderFields, stageConfigurer.getFailureCollector());
+    stageConfigurer.getFailureCollector().getOrThrowException();
+    stageConfigurer.setOutputSchema(getOutputSchema(inputSchema, aggregates));
+  }
+
+  public void validate(Schema inputSchema, List<String> partitionFields,
+                       List<WindowAggregationConfig.FunctionInfo> aggregates, List<String> partitionOrderFields,
+                       FailureCollector collector) {
+
+    for (String partitionField : partitionFields) {
+      Schema.Field field = inputSchema.getField(partitionField);
+      if (field == null) {
+        collector.addFailure(
+          String.format("Can not partition by field '%s' because it does not exist in input schema.", partitionField),
```

Comment:
We can just say `Partition field '%s' must exist in input schema.` and keep the corrective action blank